File: P2.py
# -*- coding: utf-8 -*-
"""
Created on Mon Apr  5 16:02:23 2021

@author: jgdch
"""

#Space Complexity: O(1)
#Time Complexity: O(MN), where M is the number of columns , N is the number of rows
class Solution:
    def findDiagonalOrder(self, matrix: List[List[int]]) -> List[int]:
        result=[]
        if not matrix:
            return []
        n=len(matrix)
        m=len(matrix[0])
        row,col=0,0
        direc=1
        

        while row<n and col<m:
            
             result.append(matrix[row][col])
             if direc==1:
                mrow=row-1
             else:
                mrow=row+1
             if direc==1:
                mcol=col+1
             else:
                mcol=col-1
             if mrow < 0 or mrow == n or mcol < 0 or mcol == m:
                
              
                if direc:
                    
                   if col==m-1:
                       row+=1
                   else:
                     row=row
                   if col<m-1:
                       col+=1
                   else:
                     col=col
                    
                    
                else:
                    if row==n-1:
                       col+=1
                    else:
                      col=col
                    if row<n-1:
                       row+=1
                    else:
                     row=row
                    
                
                direc= 1 - direc
             else:
                row=mrow
                col=mcol
        return result

File: p2.java
// Time Complexity : O(m*n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
//Traversing through the matrix in the way that is shown in the question. Using a boolean falg to determin if we have to go right or left diagonally. Handling edge cases using if and else

// Your code here along with comments explaining your approach

class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        int m = mat.length;
        int n = mat[0].length;
        // Output array
        int[] result = new int[(m * n)];
        // Flag user to determine whether to go right or left
        Boolean flag = true;
        int i = 0;
        int j = 0;

        for (int k = 0; k < result.length; k++) {
            result[k] = mat[i][j];
            // Conditions for going right
            if (flag) {
                if (i == 0 && j == n - 1) {

                    i++;
                    flag = false;
                } else if (i == 0) {
                    j++;

                    flag = false;
                } else if (j == n - 1) {
                    i++;

                    flag = false;
                } else {

                    i--;
                    j++;
                }
            }

            // Conditions for going left
            else {
                if (j == 0 && i == m - 1) {

                    j++;
                    flag = true;
                } else if (j == 0) {

                    i++;
                    flag = true;
                } else if (i == m - 1) {
                    j++;

                    flag = true;
                } else {
                    j--;
                    i++;
                }
            }
        }

        return result;
    }
}